File: amr/core/geometry.py
"""Geometry primitives shared by all subsystems."""
from __future__ import annotations

import math
from typing import List, Tuple

import numpy as np

from amr.core.types import Pose2D


def wrap_angle(a: float) -> float:
    """Wrap a scalar angle to (-pi, pi]."""
    a = math.fmod(a + math.pi, 2.0 * math.pi)
    if a <= 0.0:
        a += 2.0 * math.pi
    return a - math.pi


def wrap_angles(a: np.ndarray) -> np.ndarray:
    """Vector wrap to [-pi, pi) — note the half-open end differs from wrap_angle."""
    return (np.asarray(a) + np.pi) % (2.0 * np.pi) - np.pi


def pose_compose(a: Pose2D, b: Pose2D) -> Pose2D:
    """a ⊕ b: pose of frame b (expressed in a) in a's parent frame."""
    c, s = math.cos(a.theta), math.sin(a.theta)
    return Pose2D(a.x + c * b.x - s * b.y,
                  a.y + s * b.x + c * b.y,
                  wrap_angle(a.theta + b.theta))


def pose_between(a: Pose2D, b: Pose2D) -> Pose2D:
    """a ⊖ b: pose of b expressed in frame a, so pose_compose(a, result) == b."""
    dx, dy = b.x - a.x, b.y - a.y
    c, s = math.cos(a.theta), math.sin(a.theta)
    return Pose2D(c * dx + s * dy, -s * dx + c * dy, wrap_angle(b.theta - a.theta))


def transform_points(pose: Pose2D, pts: np.ndarray) -> np.ndarray:
    """Transform (N, 2) points from pose's frame into the world frame."""
    c, s = math.cos(pose.theta), math.sin(pose.theta)
    rot = np.array([[c, -s], [s, c]])
    return pts @ rot.T + np.array([pose.x, pose.y])


def bresenham(r0: int, c0: int, r1: int, c1: int) -> List[Tuple[int, int]]:
    """All integer grid cells on the line from (r0,c0) to (r1,c1), inclusive."""
    cells = []
    dr, dc = abs(r1 - r0), abs(c1 - c0)
    sr = 1 if r1 >= r0 else -1
    sc = 1 if c1 >= c0 else -1
    err = dc - dr
    r, c = r0, c0
    while True:
        cells.append((r, c))
        if r == r1 and c == c1:
            break
        e2 = 2 * err
        if e2 > -dr:
            err -= dr
            c += sc
        if e2 < dc:
            err += dc
            r += sr
    return cells


def distance_field(occupied: np.ndarray, resolution: float,
                   max_dist: float) -> np.ndarray:
    """Chamfer distance (m) from each cell to the nearest occupied cell.

    Iterative 8-neighbour relaxation (Bellman-Ford style), fully vectorized;
    converges in <= ceil(max_dist/resolution) sweeps. No scipy required.
    """
    big = max_dist / resolution + 2.0
    d = np.where(occupied, 0.0, big)
    sq2 = math.sqrt(2.0)
    for _ in range(int(math.ceil(max_dist / resolution)) + 1):
        nd = d.copy()
        nd[1:, :] = np.minimum(nd[1:, :], d[:-1, :] + 1.0)
        nd[:-1, :] = np.minimum(nd[:-1, :], d[1:, :] + 1.0)
        nd[:, 1:] = np.minimum(nd[:, 1:], d[:, :-1] + 1.0)
        nd[:, :-1] = np.minimum(nd[:, :-1], d[:, 1:] + 1.0)
        nd[1:, 1:] = np.minimum(nd[1:, 1:], d[:-1, :-1] + sq2)
        nd[1:, :-1] = np.minimum(nd[1:, :-1], d[:-1, 1:] + sq2)
        nd[:-1, 1:] = np.minimum(nd[:-1, 1:], d[1:, :-1] + sq2)
        nd[:-1, :-1] = np.minimum(nd[:-1, :-1], d[1:, 1:] + sq2)
        if np.array_equal(nd, d):
            break
        d = nd
    return np.minimum(d * resolution, max_dist).astype(np.float32)

File: src/amr_core/src/geometry.cpp
// SPDX-License-Identifier: Apache-2.0
#include "amr_core/geometry.hpp"

#include <algorithm>
#include <cmath>

namespace amr_core {

double wrap_angle(double a) {
  a = std::fmod(a + M_PI, 2.0 * M_PI);
  if (a <= 0.0) {
    a += 2.0 * M_PI;
  }
  return a - M_PI;
}

Pose2D pose_compose(const Pose2D& a, const Pose2D& b) {
  const double c = std::cos(a.theta);
  const double s = std::sin(a.theta);
  return Pose2D{a.x + c * b.x - s * b.y, a.y + s * b.x + c * b.y,
                wrap_angle(a.theta + b.theta)};
}

Pose2D pose_between(const Pose2D& a, const Pose2D& b) {
  const double dx = b.x - a.x;
  const double dy = b.y - a.y;
  const double c = std::cos(a.theta);
  const double s = std::sin(a.theta);
  return Pose2D{c * dx + s * dy, -s * dx + c * dy, wrap_angle(b.theta - a.theta)};
}

std::vector<double> transform_points(const Pose2D& pose,
                                     const std::vector<double>& pts) {
  const double c = std::cos(pose.theta);
  const double s = std::sin(pose.theta);
  std::vector<double> out(pts.size());
  for (std::size_t i = 0; i + 1 < pts.size(); i += 2) {
    const double x = pts[i];
    const double y = pts[i + 1];
    out[i] = c * x - s * y + pose.x;
    out[i + 1] = s * x + c * y + pose.y;
  }
  return out;
}

std::vector<std::pair<int, int>> bresenham(int r0, int c0, int r1, int c1) {
  std::vector<std::pair<int, int>> cells;
  const int dr = std::abs(r1 - r0);
  const int dc = std::abs(c1 - c0);
  const int sr = (r1 >= r0) ? 1 : -1;
  const int sc = (c1 >= c0) ? 1 : -1;
  int err = dc - dr;
  int r = r0;
  int c = c0;
  while (true) {
    cells.emplace_back(r, c);
    if (r == r1 && c == c1) {
      break;
    }
    const int e2 = 2 * err;
    if (e2 > -dr) {
      err -= dr;
      c += sc;
    }
    if (e2 < dc) {
      err += dc;
      r += sr;
    }
  }
  return cells;
}

std::vector<float> distance_field(const std::vector<uint8_t>& occupied,
                                  int rows, int cols, double resolution,
                                  double max_dist) {
  const float big = static_cast<float>(max_dist / resolution + 2.0);
  const float sq2 = static_cast<float>(std::sqrt(2.0));
  const std::size_t n = static_cast<std::size_t>(rows) * cols;

  std::vector<float> d(n);
  for (std::size_t i = 0; i < n; ++i) {
    d[i] = occupied[i] ? 0.0f : big;
  }

  const int sweeps = static_cast<int>(std::ceil(max_dist / resolution)) + 1;
  auto idx = [cols](int r, int c) {
    return static_cast<std::size_t>(r) * cols + c;
  };

  std::vector<float> nd(n);
  for (int s = 0; s < sweeps; ++s) {
    nd = d;
    for (int r = 0; r < rows; ++r) {
      for (int c = 0; c < cols; ++c) {
        float best = d[idx(r, c)];
        if (r > 0) best = std::min(best, d[idx(r - 1, c)] + 1.0f);
        if (r < rows - 1) best = std::min(best, d[idx(r + 1, c)] + 1.0f);
        if (c > 0) best = std::min(best, d[idx(r, c - 1)] + 1.0f);
        if (c < cols - 1) best = std::min(best, d[idx(r, c + 1)] + 1.0f);
        if (r > 0 && c > 0) best = std::min(best, d[idx(r - 1, c - 1)] + sq2);
        if (r > 0 && c < cols - 1)
          best = std::min(best, d[idx(r - 1, c + 1)] + sq2);
        if (r < rows - 1 && c > 0)
          best = std::min(best, d[idx(r + 1, c - 1)] + sq2);
        if (r < rows - 1 && c < cols - 1)
          best = std::min(best, d[idx(r + 1, c + 1)] + sq2);
        nd[idx(r, c)] = best;
      }
    }
    if (nd == d) {
      break;
    }
    d.swap(nd);
  }

  std::vector<float> out(n);
  for (std::size_t i = 0; i < n; ++i) {
    out[i] = static_cast<float>(
        std::min(static_cast<double>(d[i]) * resolution, max_dist));
  }
  return out;
}

}  // namespace amr_core
